httpbakery/agent/agent.go

```diff
@@ -0,0 +1,161 @@
+/*
+Package agent implements the agent login mechanism. The agent login
+mechanism provides a form of interaction that can be used in
+situations where interaction with the user is not possible.
+
+An agent login works as follows:
+
+	    Agent                            Login Service
+	      |                                    |
+	      | GET visitURL with agent cookie     |
+	      |----------------------------------->|
+	      |                                    |
+	      |    Macaroon with local third-party | 
+	      |                             caveat |
+	      |<-----------------------------------|
+	      |                                    |
+	      | GET visitURL with agent cookie &   |
+	      | discharged macaroon                |
+	      |----------------------------------->|
+	      |                                    |
+	      |               Agent login response |
+	      |<-----------------------------------|
+	      |                                    |
+
+The agent cookie is a cookie named "agent-login" its value is a base64
+encoded JSON object of the form:
+
+	{
+		"username": <username>,
+		"public_key": <public key>
+	}
+
+A local third-party caveat is a third party caveat with the location
+set to "local" and the caveat encrypted with the public key declared
+in the agent cookie. The httpbakery.Client automatically discharges
+the local third-party caveat.
+
+The agent login response will be a JSON object of the same form as the
+agent cookie. It will most likely contain the data specified in the
+original agent cookie. This is used to determing that a successful
+status if for an agent login and not, for example, successfully
+accessing an interactive login page.
+*/
+package agent
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"io/ioutil"
+	"mime"
+	"net/http"
+	"net/url"
+
+	"github.com/juju/loggo"
+	"gopkg.in/errgo.v1"
+
+	"gopkg.in/macaroon-bakery.v1/bakery"
+	"gopkg.in/macaroon-bakery.v1/httpbakery"
+)
+
+var logger = loggo.GetLogger("httpbakery.agent")
+
+// agentLogin defines the structure of an agent login cooke. It is also
+// returned in a successful agent login attempt to help indicate that an
+// agent login has occurred.
+type agentLogin struct {
+	Username  string            `json:"username"`
+	PublicKey *bakery.PublicKey `json:"public_key"`
+}
+
+// SetCookie creates a cookie in c which is suitable for performing agent
+// logins to u. The username will be combined with the public key taken
+// from c.Key.Public, c.Key must have been set before calling SetCookie.
+// If c is used with VisitWebPage interactions with u will be performed as
+// an agent.
+func SetCookie(c *httpbakery.Client, u *url.URL, username string) {
+	al := agentLogin{
+		Username:  username,
+		PublicKey: &c.Key.Public,
```

Comment:
Alternatively make the args to SetCookie a little closer
to what it actually needs:

    func SetCookie(u *url.URL, username string, key *bakery.PublicKey)

then it's obvious to the caller that the public key is required,
and perhaps the role of the function is a little clearer too.